vck-initializer/vck-initializer/main.go

```diff
@@ -0,0 +1,256 @@
+}
+
+type config struct {
+	Containers []corev1.Container
+	Volumes    []corev1.Volume
+}
+
+func main() {
+	flag.StringVar(&annotation, "annotation", defaultAnnotation, "The annotation to trigger initialization")
+	flag.StringVar(&initializerName, "initializer-name", defaultInitializerName, "The initializer name")
+	flag.StringVar(&namespace, "namespace", defaultNamespace, "The configuration namespace")
+	flag.BoolVar(&requireAnnotation, "require-annotation", true, "Require annotation for initialization")
+	flag.Parse()
+
+	log.Println("Starting the Kubernetes initializer...")
+	log.Printf("Initializer name set to: %s", initializerName)
+
+	clusterConfig, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	clientset, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	crdClient, err := vckv1_client.NewForConfig(clusterConfig)
+	if err != nil {
+		panic(err)
+	}
+
+	// Watch uninitialized Deployments in all namespaces.
+	restClient := clientset.AppsV1beta1().RESTClient()
+	watchlist := cache.NewListWatchFromClient(restClient, "deployments", corev1.NamespaceAll, fields.Everything())
+
+	// Wrap the returned watchlist to workaround the inability to include
+	// the `IncludeUninitialized` list option when setting up watch clients.
+	includeUninitializedWatchlist := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.IncludeUninitialized = true
+			return watchlist.List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.IncludeUninitialized = true
+			return watchlist.Watch(options)
+		},
+	}
+
+	resyncPeriod := 30 * time.Second
+
+	_, controller := cache.NewInformer(includeUninitializedWatchlist, &v1beta1.Deployment{}, resyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				o := obj.(*v1beta1.Deployment)
+				err := initializeDeployment(o, clientset, crdClient)
+				if err != nil {
+					log.Println(err)
+					log.Println("Deleteting Deployment " + o.Name)
+					deletePolicy := metav1.DeletePropagationBackground
+					err := clientset.AppsV1().Deployments(o.Namespace).Delete(o.Name, &metav1.DeleteOptions{
+						PropagationPolicy: &deletePolicy,
+					})
+					if err != nil {
+						panic(err)
+					}
+					log.Println("Deleted Deployment.")
+
+				}
+			},
+		},
+	)
+
+	stop := make(chan struct{})
+	go controller.Run(stop)
+
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+	<-signalChan
+
+	log.Println("Shutdown signal received, exiting...")
+	close(stop)
+}
+
+func initializeDeployment(deployment *v1beta1.Deployment, clientset *kubernetes.Clientset, crdClient *vckv1_client.Clientset) error {
+	if deployment.ObjectMeta.GetInitializers() != nil {
+		pendingInitializers := deployment.ObjectMeta.GetInitializers().Pending
+
+		if initializerName == pendingInitializers[0].Name {
```

Comment:
Is it always the case that the first pending initializer is always going to be this initializer?

Reply:
yes, I think the controller waits till it reaches first in the queue